File: LintCode 31. Partition Array.js
/**
 * @param nums: The integer array you should partition
 * @param k: An integer
 * @return: The index after partition
 */
const partitionArray = function (nums, k) {
    if ( nums === null || nums.length === 0 ) {
        return 0;
    }

    var left = 0;
    var right = nums.length - 1;

    while ( left <= right ) {
        while ( left <= right && nums[left] < k ) {
            left++;
        }

        while ( left <= right && nums[right] >= k ) {
            right--;
        }

        if ( left <= right ) {
            var tmp = nums[left];
            nums[left] = nums[right];
            nums[right] = tmp;

            left++;
            right--;
        }
    }

    return left;
};

File: LintCode 31. Partition Array.py
class Solution:
    """
    @param nums: The integer array you should partition
    @param k: An integer
    @return: The index after partition
    """

    def partitionArray(self, nums, k):
        lo, hi = 0, len(nums) - 1

        while lo <= hi:
            while lo <= hi and nums[lo] < k:
                lo += 1
            while lo <= hi and nums[hi] >= k:
                hi -= 1
            if lo <= hi:
                nums[lo], nums[hi] = nums[hi], nums[lo]
                lo += 1
                hi -= 1
        return lo
